File: cyclosynth/cliffords.py
"""A module that defines Clifford gates and operations involving them."""

from cyclosynth.algebra import DyadicComplexNumber
from cyclosynth.algebra import RingRoot2
from cyclosynth.ratio import AlgebraicIntegerOverRoot2
from cyclosynth.matrix import SO3Matrix
from cyclosynth.matrix import U2Matrix


zero = DyadicComplexNumber([0, 0, 0, 0, 0, 0, 0, 0], 0)
one = DyadicComplexNumber([1, 0, 0, 0, 0, 0, 0, 0], 0)
half_one = DyadicComplexNumber([1, 0, 0, 0, 0, 0, 0, 0], 1)
imag = DyadicComplexNumber([0, 0, 0, 0, 1, 0, 0, 0], 0)
half_imag = DyadicComplexNumber([0, 0, 0, 0, 1, 0, 0, 0], 1)
one_over_sqrt2 = DyadicComplexNumber([0, 0, 1, 0, 0, 0 , -1, 0], 1)


I = U2Matrix([one, zero, zero, one])
H = U2Matrix([one_over_sqrt2, one_over_sqrt2, one_over_sqrt2, -one_over_sqrt2])
S = U2Matrix([one, zero, zero, imag])
X = U2Matrix([zero, one, one, zero])
Y = U2Matrix([zero, -imag, imag, zero])
Z = U2Matrix([one, zero, zero, -one])
HX = H * X
HY = H * Y
HZ = H * Z
SX = S * X
SY = S * Y
SZ = S * Z
HS = H * S
HSX = H * S * X
HSY = H * S * Y
HSZ = H * S * Z
SH = S * H
SHX = S * H * X
SHY = S * H * Y
SHZ = S * H * Z
HSH = H * S * H
HSHX = H * S * H * X
HSHY = H * S * H * Y
HSHZ = H * S * H * Z
H_dg = H
S_dg = S * S * S
X_dg = X
Y_dg = Y
Z_dg = Z
HX_dg = X * H
HY_dg = Y * H
HZ_dg = Z * H
SX_dg = X * S_dg
SY_dg = Y * S_dg
SZ_dg = Z * S_dg
HS_dg = S_dg * H
HSX_dg = X * S_dg * H
HSY_dg = Y * S_dg * H
HSZ_dg = Z * S_dg * H
SH_dg = H * S_dg
SHX_dg = X * H * S_dg
SHY_dg = Y * H * S_dg
SHZ_dg = Z * H * S_dg
HSH_dg = H * S_dg * H
HSHX_dg = X * H * S_dg * H
HSHY_dg = Y * H * S_dg * H
HSHZ_dg = Z * H * S_dg * H


clifford_gates_to_u2 = {
    'I': I,
    'H': H,
    'S': S,
    'X': X,
    'Y': Y,
    'Z': Z,
    'XH': HX,
    'YH': HY, 
    'ZH': HZ,
    'XS': SX,
    'YS': SY,
    'ZS': SZ,
    'SH': HS,
    'XSH': HSX,
    'YSH': HSY,
    'ZSH': HSZ,
    'HS': SH,
    'XHS': SHX,
    'YHS': SHY,
    'ZHS': SHZ,
    'HSH': HSH,
    'XHSH': HSHX,
    'YHSH': HSHY,
    'ZHSH': HSHZ,
}


clifford_gates_to_invu2 = {
    'I': I,
    'H': H,
    'S': S_dg,
    'X': X,
    'Y': Y,
    'Z': Z,
}


def cliffords(as_u2: bool = False) -> list[U2Matrix | SO3Matrix]:
    """
    Returns a list of single qubit Clifford group elements.

    Args:
        as_u2 (bool): If True, return the Clifford gates as U(2) matrices.
            Otherwise, return them as SO(3) matrices.

    Note:
        Clifford group is all C = AB
        A in {I, H, S, HS, SH, HSH}
        B in {I, X, Y, Z}
    """
    if as_u2:
        cliffords = [
            I, H, S, X, Y, Z, HX, HY, HZ, SX, SY, SZ, HS, HSX,
            HSY, HSZ, SH, SHX, SHY, SHZ, HSH, HSHX, HSHY, HSHZ,
        ]
    else:
        p = AlgebraicIntegerOverRoot2(RingRoot2([1, 0]), 0)
        m = AlgebraicIntegerOverRoot2(RingRoot2([-1, 0]), 0)
        z = AlgebraicIntegerOverRoot2(RingRoot2([0, 0]), 0)
        cliffords = [
            SO3Matrix([p, z, z, z, p, z, z, z, p]),
            SO3Matrix([z, z, p, z, m, z, p, z, z]),
            SO3Matrix([z, p, z, m, z, z, z, z, p]),
            SO3Matrix([p, z, z, z, m, z, z, z, m]),
            SO3Matrix([m, z, z, z, p, z, z, z, m]),
            SO3Matrix([m, z, z, z, m, z, z, z, p]),
            SO3Matrix([z, z, p, z, p, z, m, z, z]),
            SO3Matrix([z, z, m, z, m, z, m, z, z]),
            SO3Matrix([z, z, m, z, p, z, p, z, z]),
            SO3Matrix([z, p, z, p, z, z, z, z, m]),
            SO3Matrix([z, m, z, m, z, z, z, z, m]),
            SO3Matrix([z, m, z, p, z, z, z, z, p]),
            SO3Matrix([z, m, z, z, z, m, p, z, z]),
            SO3Matrix([z, m, z, z, z, p, m, z, z]),
            SO3Matrix([z, p, z, z, z, m, m, z, z]),
            SO3Matrix([z, p, z, z, z, p, p, z, z]),
            SO3Matrix([z, z, p, p, z, z, z, p, z]),
            SO3Matrix([z, z, p, m, z, z, z, m, z]),
            SO3Matrix([z, z, m, p, z, z, z, m, z]),
            SO3Matrix([z, z, m, m, z, z, z, p, z]),
            SO3Matrix([p, z, z, z, z, p, z, m, z]),
            SO3Matrix([p, z, z, z, z, m, z, p, z]),
            SO3Matrix([m, z, z, z, z, p, z, p, z]),
            SO3Matrix([m, z, z, z, z, m, z, m, z]),
        ]
    return cliffords


def match_clifford(matrix: U2Matrix) -> str | None:
    """
    Given a matrix in U(2), return the corresponding Clifford gates.

    If the matrix is not a Clifford gate, None will be returned.
    """
    candidates = [
        gates for gates, u in clifford_gates_to_u2.items()
        if matrix.hilbert_schmidt_distance(u) <= 1e-8
    ]
    if len(candidates) == 0:
        return None
    else:
        return candidates[0]


def clifford_str_to_u2(gates: str) -> U2Matrix | None:
    return clifford_gates_to_u2.get(gates)

File: cyclosynth-rs/src/bloch/cliffords.rs
//! The 24 single-qubit Clifford group elements as U2 matrices, and Clifford matching.

use crate::algebra::DyadicComplexNumber;
use crate::matrix::U2Matrix;

fn dcn(vals: [i128; 8], exp: i32) -> DyadicComplexNumber {
    DyadicComplexNumber::new(vals.to_vec(), exp)
}

fn zero() -> DyadicComplexNumber { dcn([0,0,0,0,0,0,0,0], 0) }
fn one()  -> DyadicComplexNumber { dcn([1,0,0,0,0,0,0,0], 0) }
fn neg_one() -> DyadicComplexNumber { dcn([-1,0,0,0,0,0,0,0], 0) }
fn imag() -> DyadicComplexNumber { dcn([0,0,0,0,1,0,0,0], 0) }
fn neg_imag() -> DyadicComplexNumber { dcn([0,0,0,0,-1,0,0,0], 0) }
fn one_over_sqrt2() -> DyadicComplexNumber { dcn([0,0,1,0,0,0,-1,0], 1) }
fn neg_one_over_sqrt2() -> DyadicComplexNumber { dcn([0,0,-1,0,0,0,1,0], 1) }

/// Build the 24 single-qubit Clifford U2 matrices, each paired with its gate string.
///
/// Gate string convention matches the Python `clifford_gates_to_u2` dict.
pub fn clifford_table() -> Vec<(&'static str, U2Matrix)> {
    let i_mat  = U2Matrix::new([one(), zero(), zero(), one()]);
    let h      = U2Matrix::new([one_over_sqrt2(), one_over_sqrt2(),
                                one_over_sqrt2(), neg_one_over_sqrt2()]);
    let s      = U2Matrix::new([one(), zero(), zero(), imag()]);
    let x      = U2Matrix::new([zero(), one(), one(), zero()]);
    let y      = U2Matrix::new([zero(), neg_imag(), imag(), zero()]);
    let z      = U2Matrix::new([one(), zero(), zero(), neg_one()]);

    let hx  = h.mul(&x);
    let hy  = h.mul(&y);
    let hz  = h.mul(&z);
    let sx  = s.mul(&x);
    let sy  = s.mul(&y);
    let sz  = s.mul(&z);
    let hs  = h.mul(&s);
    let hsx = h.mul(&s).mul(&x);
    let hsy = h.mul(&s).mul(&y);
    let hsz = h.mul(&s).mul(&z);
    let sh  = s.mul(&h);
    let shx = s.mul(&h).mul(&x);
    let shy = s.mul(&h).mul(&y);
    let shz = s.mul(&h).mul(&z);
    let hsh  = h.mul(&s).mul(&h);
    let hshx = h.mul(&s).mul(&h).mul(&x);
    let hshy = h.mul(&s).mul(&h).mul(&y);
    let hshz = h.mul(&s).mul(&h).mul(&z);

    vec![
        ("I",    i_mat),
        ("H",    h),
        ("S",    s),
        ("X",    x),
        ("Y",    y),
        ("Z",    z),
        ("XH",   hx),
        ("YH",   hy),
        ("ZH",   hz),
        ("XS",   sx),
        ("YS",   sy),
        ("ZS",   sz),
        ("SH",   hs),
        ("XSH",  hsx),
        ("YSH",  hsy),
        ("ZSH",  hsz),
        ("HS",   sh),
        ("XHS",  shx),
        ("YHS",  shy),
        ("ZHS",  shz),
        ("HSH",  hsh),
        ("XHSH", hshx),
        ("YHSH", hshy),
        ("ZHSH", hshz),
    ]
}

/// Match `matrix` to a single-qubit Clifford gate, returning its gate string.
///
/// Returns `None` if the matrix is not within 1e-8 Hilbert-Schmidt distance of
/// any Clifford.  Returns `None` also for the identity ("I") since appending
/// nothing is the right behavior.
pub fn match_clifford(matrix: &U2Matrix) -> Option<String> {
    for (name, clifford) in clifford_table() {
        if matrix.hilbert_schmidt_distance(&clifford) <= 1e-8 {
            if name == "I" {
                return None;
            }
            return Some(name.to_string());
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::matrix::factories::unitary_identity;

    #[test]
    fn test_clifford_table_size() {
        assert_eq!(clifford_table().len(), 24);
    }

    #[test]
    fn test_identity_matches_none() {
        let id = unitary_identity(4);
        assert!(match_clifford(&id).is_none());
    }

    #[test]
    fn test_h_matches() {
        let h = U2Matrix::new([
            one_over_sqrt2(), one_over_sqrt2(),
            one_over_sqrt2(), neg_one_over_sqrt2(),
        ]);
        assert_eq!(match_clifford(&h), Some("H".to_string()));
    }

    #[test]
    fn test_all_cliffords_self_match() {
        for (name, mat) in clifford_table() {
            let result = match_clifford(&mat);
            if name == "I" {
                assert!(result.is_none(), "identity should return None");
            } else {
                assert_eq!(result, Some(name.to_string()), "gate {name} should match itself");
            }
        }
    }
}
